feat(Firestore): Add support for FirestoreChatMessageHistory (#60)

KennethKnudsen97 opened the pull request:
- Description: This adds support for Firestore chat message history. With this feature, you can use firestore as the memory for your AI agent.
  - Issue: Resolves #60 .



- [x] Create example on how to use Firestore chat message history
- [ ] Fill out TODO in Readme.md

davidmigloz commented:
Hey @KennethKnudsen97, thank you very much for the PR!
I'll review it this weekend.

davidmigloz commented on packages/langchain/lib/src/model_io/chat_models/models/models.dart:
```diff
@@ -337,6 +400,15 @@ CustomChatMessage{
+
+  @override
+  Map<String, dynamic> toJson() => {
+        'type': 'CustomChatMessage',
```
Let's use `'type': 'custom'` for this one.

davidmigloz commented on packages/langchain_firestore/pubspec.yaml:
```diff
@@ -0,0 +1,17 @@
+name: langchain_firestore
```
The current modularization strategy I follow is creating a package per company. In the case of Firebase, it's owned by Google so `FirestoreChatMessageHistory` can go in the existing `langchain_google` package. 

Let's move it to:
```
langchain_google / lib / stores / firestore.dart
```

davidmigloz commented on packages/langchain_firestore/lib/src/langchain_firestore_base.dart:
```diff
@@ -0,0 +1,126 @@
+      return FirestoreChatMessageField.fromJson(doc.data());
+    }).toList();
+
+    return conversation.map((e) => e.message).toList();
```
We can avoid this second map if we take the message already in the previous map. In this way, we avoid iterating over all the messages again.

davidmigloz commented on packages/langchain_firestore/lib/src/langchain_firestore_base.dart:
```diff
@@ -0,0 +1,126 @@
+
+  @override
+  Future<void> addChatMessage(ChatMessage message) async {
+    print("adding message to firebase");
```
Let's remove this print, I guess you used it for debugging

davidmigloz commented on packages/langchain_firestore/lib/src/langchain_firestore_base.dart:
```diff
@@ -0,0 +1,126 @@
+import 'package:langchain/langchain.dart';
+
+final class FirestoreChatMessageHistory extends BaseChatMessageHistory {
+  FirestoreChatMessageHistory(this.collection);
```
Maybe instead of passing the `CollectionReference`, we can have the following constructor parameters:
- `firestoreClient`
- `collectionName`
- `sessionId`
- `userId`

So then internally we can get the document like:
```
firestoreClient
      .collection(this.collectionName)
      .doc(this.sessionId);
```

Take a look to the LangChain JS implementation:
https://github.com/langchain-ai/langchainjs/blob/main/langchain/src/stores/message/firestore.ts

KennethKnudsen97 replied:
I was thinking about this too, but I couldn't see how it would work if you had nested collections
Something like this: 
```
      FirebaseFirestore.instance
          .collection("Users")
          .doc(Auth().currentUser!.uid)
          .collection("Langchain");

```
or if you had an even deeper nested collection.


let me know what you think.

KennethKnudsen97 replied:
That is properly the correct solution :+1:  
I will fix this when I get some time again.

davidmigloz replied:
awesome!

davidmigloz commented:
Hey @KennethKnudsen97, 

I reviewed the PR and added some comments, take a look at them and let me know if they make sense to you. 
Thanks for your work again!

mdshihabuddinroky commented:
did you add a retrieval parameter to add retrieval?

davidmigloz commented:
hey @mdshihabuddinroky,

Can you explain more about what you mean?

KennethKnudsen97 commented:
Made a huge mess syncing my branch, so i had to do some clean up..
Its my first time forking a project, so i will learn :D

davidmigloz commented:
No worries! Let me know if you need help 🙂

KennethKnudsen97 commented:
I finally got some time to change the arguments for FirestoreChatMessageHistory. 
Now they take in a collection string which can be a nested collection as well. 
I use it in my program like this. 
 ```
var user = Auth().currentUser!.uid;
    history = FirestoreChatMessageHistory(
      collections: "Users/$user/Langchain",
      options: widget.firebaseOptions,
      userId: user,
    );
```
Because we pass in firebaseoptions we now have to initialize firebase. 
by calling: 
`history.ensureFirestore();`


I also added a userId field in the argument, so that every message created has a userId.

What do you think about that @davidmigloz

davidmigloz commented:
Hey @KennethKnudsen97!

Sorry for the late response.. and thanks again for your work 🙂 

The new arguments look good. I would remove the Firebase App initialization to the user (similar to [how `ChatFirebaseVertexAI` does](https://langchaindart.com/#/modules/model_io/models/chat_models/integrations/firebase_vertex_ai?id=_3-initialize-your-firebase-app)), in that way we don't need the `ensureFirestore` method.

As you may have seen, we now have a specific `langchain_firebase` package, so we should move this code to that package.

I have on my list to rework [how memory integrates with LCEL](https://github.com/davidmigloz/langchain_dart/issues/387), so it'd great to finally release this together with that.

KennethKnudsen97 commented:
Hey @davidmigloz 

I have now moved FirestoreChatMessageHistory into langchain_firebase. 
I have also removed the ensureInitialized function and moved the initializing of the Firebase app to the user. 
The examples for FirestoreChatMessageHistory are still in langchain_google/example is that fine?
